### PR TITLE
owmosaic: fix VizRank crashing or freezing

### PR DESCRIPTION
**Pull request by markotoplak**
The old code added some GUI manipulation into the thread. To prevent this in
the future I moved the computation loop into the worker class.

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Orange/widgets/visualize/utils/__init__.py [hunk 1]
```diff
@@ -351,23 +327,53 @@ def toggle(self):
...
+                    pos = bisect_left(self.obj.scores, score)
+                    row_items = self.obj.row_for_state(score, state)
+                    if self.obj._has_bars:
+                        bar = self.obj.bar_length(score)
+                        if bar is not None:
+                            row_items[0].setData(bar,
+                                                 gui.TableBarItem.BarRole)
+                    self.obj.rank_model.insertRow(pos, row_items)
+                    self.obj.scores.insert(pos, score)
+            except IOError:
```
Comment (kernc) [inline Orange/widgets/visualize/utils/__init__.py]: Can this catching be more precisely positioned? Where does the error happen?

Comment (markotoplak) [inline Orange/widgets/visualize/utils/__init__.py]: Whoops. I did this for debugging so I could actually see the exception and put here something that could not happen at all. :) I'll change it into an undefined exception as it was! Thanks!

Comment (kernc) [inline Orange/widgets/visualize/utils/__init__.py]: Or, instead, change it into Exception **and** move it above right below the `compute_score()` line. 😄

File: Orange/widgets/visualize/utils/__init__.py [hunk 2]
```diff
@@ -351,23 +327,52 @@ def toggle(self):
...
+            try:
+                score = self.obj.compute_score(state)
+                if score is not None:
+                    pos = bisect_left(self.obj.scores, score)
+                    row_items = self.obj.row_for_state(score, state)
+                    bar_length = self.obj.bar_length(score)
+                    if bar_length is not None:
+                        row_items[0].setData(bar_length,
+                                             gui.TableBarItem.BarRole)
+                    self.obj.rank_model.insertRow(pos, row_items)
```
Comment (ales-erjavec) [inline Orange/widgets/visualize/utils/__init__.py]: This is modifying a model while it is being observed from a main thread by both a view and a QSortFilterProxyModel. This is not thread safe.

I can produce a segfault if I run the ranking and (persistently) edit the 'Filter ...' along with some item selection in the view.
```
Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff94ffff06 __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff99f2a4ec pthread_kill + 90
2   libsystem_c.dylib             	0x00007fff9c3626df abort + 129
3   QtCore                        	0x00000001117f8e79 qt_message_fatal(QtMsgType, QMessageLogContext const&, QString const&) + 9
4   QtCore                        	0x00000001117fa847 QMessageLogger::fatal(char const*, ...) const + 231
5   QtCore.so                     	0x0000000111626044 pyqt5_err_print() + 948
6   QtCore.so                     	0x000000011162b777 PyQtSlotProxy::unislot(void**) + 103
7   QtCore.so                     	0x000000011162b6da PyQtSlotProxy::qt_metacall(QMetaObject::Call, int, void**) + 58
8   QtCore                        	0x00000001119f9d9b QMetaObject::activate(QObject*, int, int, void**) + 2027
9   QtCore                        	0x0000000111985284 QItemSelectionModelPrivate::_q_rowsAboutToBeRemoved(QModelIndex const&, int, int) + 3236
10  QtCore                        	0x00000001119f9923 QMetaObject::activate(QObject*, int, int, void**) + 883
11  QtCore                        	0x0000000111a8cb7e QAbstractItemModel::rowsAboutToBeRemoved(QModelIndex const&, int, int, QAbstractItemModel::QPrivateSignal) + 78
12  QtCore                        	0x000000011197a84f QAbstractItemModel::beginRemoveRows(QModelIndex const&, int, int) + 319
13  QtCore                        	0x000000011199583a QSortFilterProxyModelPrivate::remove_proxy_interval(QVector<int>&, QVector<int>&, int, int, QModelIndex const&, Qt::Orientation, bool) + 74
14  QtCore                        	0x0000000111995768 QSortFilterProxyModelPrivate::remove_source_items(QVector<int>&, QVector<int>&, QVector<int> const&, QModelIndex const&, Qt::Orientation, bool) + 328
15  QtCore                        	0x0000000111998678 QSortFilterProxyModelPrivate::handle_filter_changed(QVector<int>&, QVector<int>&, QModelIndex const&, Qt::Orientation) + 600
16  QtCore                        	0x0000000111997ec1 QSortFilterProxyModelPrivate::filter_changed(QModelIndex const&) + 193
17  QtCore                        	0x000000011199d837 QSortFilterProxyModel::setFilterFixedString(QString const&) + 135
18  QtCore.so                     	0x000000011159614a meth_QSortFilterProxyModel_setFilterFixedString(_object*, _object*) + 122
19  org.python.python             	0x000000010f512111 PyCFunction_Call + 276
20  org.python.python             	0x000000010f5724ab PyEval_EvalFrameEx + 22999
21  org.python.python             	0x000000010f5759fd _PyEval_EvalCodeWithName + 1886
22  org.python.python             	0x000000010f56cabf PyEval_EvalCodeEx + 78
23  org.python.python             	0x000000010f4fb760 function_call + 369
24  org.python.python             	0x000000010f4dbcb5 PyObject_Call + 101
25  org.python.python             	0x000000010f4ebeb5 method_call + 140
26  org.python.python             	0x000000010f4dbcb5 PyObject_Call + 101
27  org.python.python             	0x000000010f575243 PyEval_CallObjectWithKeywords + 165
28  QtCore.so                     	0x000000011162aeb8 PyQtSlot::call(_object*, _object*) const + 40
29  QtCore.so                     	0x000000011162adb5 PyQtSlot::invoke(void**, _object*, void*, bool) const + 341
30  QtCore.so                     	0x000000011162b768 PyQtSlotProxy::unislot(void**) + 88
31  QtCore.so                     	0x000000011162b6da PyQtSlotProxy::qt_metacall(QMetaObject::Call, int, void**) + 58
32  QtCore                        	0x00000001119f9d9b QMetaObject::activate(QObject*, int, int, void**) + 2027
33  QtWidgets                     	0x0000000112f27549 QLineEdit::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) + 1129
34  QtCore                        	0x00000001119f9923 QMetaObject::activate(QObject*, int, int, void**) + 883
35  QtWidgets                     	0x0000000112fb6b79 QWidgetLineControl::finishChange(int, bool, bool) + 537
36  QtWidgets                     	0x0000000112fbc2a3 QWidgetLineControl::processKeyEvent(QKeyEvent*) + 4499
37  QtWidgets                     	0x0000000112f24602 QLineEdit::keyPressEvent(QKeyEvent*) + 34
38  QtWidgets.so                  	0x00000001128880df sipQLineEdit::keyPressEvent(QKeyEvent*) + 111
39  QtWidgets                     	0x0000000112e19800 QWidget::event(QEvent*) + 4560
40  QtWidgets                     	0x0000000112f23ae4 QLineEdit::event(QEvent*) + 628
41  QtWidgets.so                  	0x0000000112887878 sipQLineEdit::event(QEvent*) + 104
42  QtWidgets                     	0x0000000112dd8edd QApplicationPrivate::notify_helper(QObject*, QEvent*) + 269
43  QtWidgets                     	0x0000000112ddb0db QApplication::notify(QObject*, QEvent*) + 4043
44  QtWidgets.so                  	0x0000000112918cd3 sipQApplication::notify(QObject*, QEvent*) + 115
45  QtCore                        	0x00000001119c8aa4 QCoreApplication::notifyInternal2(QObject*, QEvent*) + 164
46  QtWidgets                     	0x0000000112e3867f QWidgetWindow::event(QEvent*) + 367
47  QtWidgets                     	0x0000000112dd8edd QApplicationPrivate::notify_helper(QObject*, QEvent*) + 269
48  QtWidgets                     	0x0000000112ddb822 QApplication::notify(QObject*, QEvent*) + 5906
49  QtWidgets.so                  	0x0000000112918cd3 sipQApplication::notify(QObject*, QEvent*) + 115
50  QtCore                        	0x00000001119c8aa4 QCoreApplication::notifyInternal2(QObject*, QEvent*) + 164
51  QtGui                         	0x000000011221e373 QGuiApplicationPrivate::processKeyEvent(QWindowSystemInterfacePrivate::KeyEvent*) + 163
52  QtGui                         	0x000000011220706b QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 155
53  QtGui                         	0x000000011220406c QWindowSystemInterface::flushWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 572
54  libqcocoa.dylib               	0x000000011fff3c0f -[QNSView handleKeyEvent:eventType:] + 1167
55  libqcocoa.dylib               	0x000000011fff3d42 -[QNSView keyDown:] + 98
56  com.apple.AppKit              	0x00007fff932efb15 -[NSWindow _reallySendEvent:isDelayedEvent:] + 2108
57  com.apple.AppKit              	0x00007fff92d2e539 -[NSWindow sendEvent:] + 517
58  libqcocoa.dylib               	0x000000011ffe2b97 -[QNSPanel superSendEvent:] + 39
59  libqcocoa.dylib               	0x000000011ffe1f91 -[QNSWindowHelper handleWindowEvent:] + 721
60  com.apple.AppKit              	0x00007fff92caf16a -[NSApplication sendEvent:] + 4382
61  libqcocoa.dylib               	0x000000011fffcc3e -[QNSApplication sendEvent:] + 78
62  com.apple.AppKit              	0x00007fff92b15df2 -[NSApplication run] + 796
63  libqcocoa.dylib               	0x000000011fff953f QCocoaEventDispatcher::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 2191
64  QtCore                        	0x00000001119c4eb1 QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 401
65  QtCore                        	0x00000001119c9115 QCoreApplication::exec() + 341
66  QtWidgets.so                  	0x0000000112a0fc02 meth_QApplication_exec_(_object*, _object*) + 82
67  org.python.python             	0x000000010f512111 PyCFunction_Call + 276
68  org.python.python             	0x000000010f5724ab PyEval_EvalFrameEx + 22999
69  org.python.python             	0x000000010f5762ad fast_function + 220
70  org.python.python             	0x000000010f57232b PyEval_EvalFrameEx + 22615
71  org.python.python             	0x000000010f5759fd _PyEval_EvalCodeWithName + 1886
72  org.python.python             	0x000000010f56ca6b PyEval_EvalCode + 75
73  org.python.python             	0x000000010f56a547 builtin_exec + 524
74  org.python.python             	0x000000010f512111 PyCFunction_Call + 276
75  org.python.python             	0x000000010f5724ab PyEval_EvalFrameEx + 22999
76  org.python.python             	0x000000010f5759fd _PyEval_EvalCodeWithName + 1886
77  org.python.python             	0x000000010f57632d fast_function + 348
78  org.python.python             	0x000000010f57232b PyEval_EvalFrameEx + 22615
79  org.python.python             	0x000000010f5759fd _PyEval_EvalCodeWithName + 1886
80  org.python.python             	0x000000010f56cabf PyEval_EvalCodeEx + 78
81  org.python.python             	0x000000010f4fb760 function_call + 369
82  org.python.python             	0x000000010f4dbcb5 PyObject_Call + 101
83  org.python.python             	0x000000010f5a5f47 RunModule + 147
84  org.python.python             	0x000000010f5a594c Py_Main + 2504
85  org.python.python             	0x000000010f4cde3f 0x10f4cc000 + 7743
86  libdyld.dylib                 	0x00007fff908d45ad start + 1
```

Comment (markotoplak) [inline Orange/widgets/visualize/utils/__init__.py]: I changed it so that now the thread adds to a Queue and the main threads adds from it to the model.

**Pull request merged**